File: compiler/bench/matmul.c
// expected: 1254400
#include <stdio.h>
#define N 20
long A[N][N], B[N][N], C[N][N];
int main() {
    for (int i = 0; i < N; i++)
        for (int j = 0; j < N; j++) {
            A[i][j] = i + j;
            B[i][j] = j - i;
            C[i][j] = 0;
        }
    for (int i = 0; i < N; i++)
        for (int j = 0; j < N; j++)
            for (int k = 0; k < N; k++)
                C[i][j] += A[i][k] * B[k][j];
    long sum = 0;
    for (int i = 0; i < N; i++)
        for (int j = 0; j < N; j++)
            sum += C[i][j];
    printf("%ld\n", sum);
    return 0;
}

File: compiler/bench/matmul.fs
\ expected: 1254400

20 constant N
create A N N * cells allot
create B N N * cells allot
create C N N * cells allot

: idx ( row col -- addr ) swap N * + cells ;

: init ( -- )
  N 0 do N 0 do
    i j + A i j idx + !
    j i - B i j idx + !
  loop loop ;

: dot ( row col -- sum )
  0 N 0 do
    2 pick i idx A + @
    2 pick i swap idx B + @
    * +
  loop nip nip ;

: mult ( -- )
  N 0 do N 0 do
    i j dot C i j idx + !
  loop loop ;

: sum ( -- n )
  0 N N * 0 do C i cells + @ + loop ;

: main init mult sum . cr ;
